### PR TITLE
TEZ-4275: Use Google Guava Intern Facility

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

File: tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java [hunk 1]
```diff
@@ -866,10 +866,11 @@ protected synchronized void handle(DAGAppMasterEvent event) {
...
       containers.dagComplete(cleanupEvent.getDag());
-      TezTaskAttemptID.clearCache();
-      TezTaskID.clearCache();
-      TezVertexID.clearCache();
-      TezDAGID.clearCache();
+
+      // Hint to the JVM to clean up anything no longer needed; be ready for the next DAG
+      System.gc();
+      System.runFinalization();
+
```
Comment (belugabehr) [inline tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java:873]: I'm not sure if it is helpful to role this change into this ticket or not.  However, the `clearCache` stuff is pretty useless.  All it does is make the map entries available for garbage collection,... which they already are by design.

However, this here will hint to the JDK to actually clean up that stuff.  Hinting to the JDK at this points seems like a reasonable thing to do between DAGs, but I'm not sure if it's worthwhile.

**Comment by belugabehr** [comment]
OK @jteagles @abstractdog this PR is failing on some pre-existing find bugs error.  Kind of funny actually:


```
Null passed for non-null parameter of com.google.common.base.Strings.isNullOrEmpty(String)
```

The method has `isNull` in the name of it, of course it can accept a null value. :D

Please consider these changes as they should lower memory requirements, provide faster concurrent access, and simplify the code a bit.  In particular, please consider my thoughts here:

https://github.com/apache/tez/pull/95#discussion_r572336225

**Comment by abstractdog** [comment]
> OK @jteagles @abstractdog this PR is failing on some pre-existing find bugs error. Kind of funny actually:
> 
> ```
> Null passed for non-null parameter of com.google.common.base.Strings.isNullOrEmpty(String)
> ```
> 
> The method has `isNull` in the name of it, of course it can accept a null value. :D
> 
> Please consider these changes as they should lower memory requirements, provide faster concurrent access, and simplify the code a bit. In particular, please consider my thoughts here:
> 
> [#95 (comment)](https://github.com/apache/tez/pull/95#discussion_r572336225)

yeah, absolutely, we've also found similar isNull* madness in https://issues.apache.org/jira/browse/TEZ-4215, I believe we should merge that @jteagles, and @belugabehr  you can handle this in a similar way...if findbugs warning simply doesn't makes sense, let's try to ignore it

**Comment by belugabehr** [comment]
@abstractdog @jteagles Any thoughts on this?

**Comment by jteagles** [comment]
@belugabehr can you compare speed and memory consumption with current implementation? I have a test that can be used to measure.
https://issues.apache.org/jira/browse/TEZ-1526
https://issues.apache.org/jira/secure/attachment/12857710/TEZ-1526.memory.test.patch
mvn clean test -Dtest=TestTezIds#testIdGetInstancePerformance -pl tez-common -am

Then compare 
- tez-common/target/surefire-reports/org.apache.tez.dag.records.TestTezIds.txt
- tez-common/target/surefire-reports/org.apache.tez.dag.records.TestTezIds-output.txt

**Comment by belugabehr** [comment]
```none
== TEZ-4275 ==

TEZ-1526: heap used memory: 100000:26299128
TEZ-1526: heap used memory: 200000:12867000
TEZ-1526: heap used memory: 300000:46693400
TEZ-1526: heap used memory: 400000:76540520
TEZ-1526: heap used memory: 500000:106386920
TEZ-1526: heap used memory: 600000:40765776
TEZ-1526: heap used memory: 700000:70625496
...
TEZ-1526: heap used memory: 9300000:115818312
TEZ-1526: heap used memory: 9400000:143028440
TEZ-1526: heap used memory: 9500000:177041136
TEZ-1526: heap used memory: 9600000:204251344
TEZ-1526: heap used memory: 9700000:231461552
TEZ-1526: heap used memory: 9800000:265474312
TEZ-1526: heap used memory: 9900000:292684504
TEZ-1526: heap used memory: 10000000:326697216
```

```none
== master ==

TEZ-1526: heap used memory: 100000:29985368
TEZ-1526: heap used memory: 200000:20945736
TEZ-1526: heap used memory: 300000:54775896
TEZ-1526: heap used memory: 400000:84626224
TEZ-1526: heap used memory: 500000:15007832
TEZ-1526: heap used memory: 600000:48851936
TEZ-1526: heap used memory: 700000:78714376
...
TEZ-1526: heap used memory: 9300000:171467736
TEZ-1526: heap used memory: 9400000:198175528
TEZ-1526: heap used memory: 9500000:224883320
TEZ-1526: heap used memory: 9600000:258268056
TEZ-1526: heap used memory: 9700000:284975880
TEZ-1526: heap used memory: 9800000:318360616
TEZ-1526: heap used memory: 9900000:345068408
TEZ-1526: heap used memory: 10000000:44603528
```

```
== TEZ-4275 ==
-------------------------------------------------------------------------------
Test set: org.apache.tez.dag.records.TestTezIds
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.648 s - in org.apache.tez.dag.records.TestTezIds

== master ==

-------------------------------------------------------------------------------
Test set: org.apache.tez.dag.records.TestTezIds
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.937 s - in org.apache.tez.dag.records.TestTezIds
```

**Comment by belugabehr** [comment]
Results match my expectations.  If nothing else, Guava uses a `Map` with the cached value (as key) and a dummy value. Current implementation uses a `Map` with the cached value (as key) and a `new` value.  That'll save some memory.

**Comment by belugabehr** [comment]
Also, I suspect Guava version is better in multi-threaded environment since it locks on "regions" of the map instead of locking on the entire map (`master`).  I ran a few quick and dirty sanity checks by wrapping the proposed unit test in a thread and launching 4 of them...

master: 3.7s - 4.2s
TEZ-4275: 3.4s - 3.6s

**Comment by belugabehr** [comment]
16 Threads yielded a similar 10% improvement in throughput

**Comment by belugabehr** [comment]
```
[2021-02-10T18:17:40.551Z]     Reason | Tests

[2021-02-10T18:17:40.551Z]  FindBugs  |  module:tez-dag 

[2021-02-10T18:17:40.551Z]            |  org.apache.tez.dag.app.DAGAppMaster.handle(DAGAppMasterEvent) forces garbage collection; extremely dubious except in benchmarking code At DAGAppMaster.java:dubious except in benchmarking code At DAGAppMaster.java:[line 872] 
```

Hello, what do you think about this PR as it currently stands?

**Comment by jteagles** [comment]
If the GC isn't strictly needed can it be taken out?

**Comment by belugabehr** [comment]
@jteagles Yes. It can be.

There may be some value though to hinting that now (between DAGs) is a good time to cleanup, and I was mimicking the previous code that was attempting to cleanup these caches.  Just let me know, I don't have a feeling (or data) on it one way or another.

**Comment by belugabehr** [comment]
@jteagles What would your preference be here?

**Comment by belugabehr** [comment]
@jteagles @abstractdog OK.  I took out the GC hint and it builds without the warning.

Thanks!

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  17m 55s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   4m 25s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   9m 10s |  master passed  |
| +1 :green_heart: |  compile  |   3m 20s |  master passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |   2m 59s |  master passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  checkstyle  |   2m 51s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 56s |  master passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   2m 40s |  master passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +0 :ok: |  spotbugs  |   0m 43s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +0 :ok: |  findbugs  |   0m 41s |  tez-tools/analyzers/job-analyzer in master has 4 extant findbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m  9s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  9s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 57s |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |   1m 57s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 38s |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  javac  |   1m 38s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m  9s |  tez-common: The patch generated 0 new + 25 unchanged - 1 fixed = 25 total (was 26)  |
| +1 :green_heart: |  checkstyle  |   0m  9s |  The patch passed checkstyle in tez-runtime-internals  |
| +1 :green_heart: |  checkstyle  |   0m 14s |  tez-runtime-library: The patch generated 0 new + 30 unchanged - 1 fixed = 30 total (was 31)  |
| +1 :green_heart: |  checkstyle  |   0m 25s |  The patch passed checkstyle in tez-dag  |
| +1 :green_heart: |  checkstyle  |   0m  9s |  The patch passed checkstyle in tez-history-parser  |
| +1 :green_heart: |  checkstyle  |   0m  9s |  tez-tools/analyzers/job-analyzer: The patch generated 0 new + 18 unchanged - 2 fixed = 18 total (was 20)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   1m 32s |  the patch passed with JDK Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   1m 23s |  the patch passed with JDK Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10  |
| +1 :green_heart: |  findbugs  |   4m 32s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 29s |  tez-common in the patch passed.  |
| +1 :green_heart: |  unit  |   0m 31s |  tez-runtime-internals in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 56s |  tez-runtime-library in the patch passed.  |
| +1 :green_heart: |  unit  |   4m  8s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 11s |  tez-history-parser in the patch passed.  |
| +1 :green_heart: |  unit  |   2m 24s |  job-analyzer in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  The patch does not generate ASF License warnings.  |
|  |   |  83m 58s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-95/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/95 |
| JIRA Issue | TEZ-4275 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux dcbb343a39ac 4.15.0-128-generic #131-Ubuntu SMP Wed Dec 9 06:57:35 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 01847814d |
| Default Java | Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.11+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_292-8u292-b10-0ubuntu1~20.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-95/6/testReport/ |
| Max. process+thread count | 789 (vs. ulimit of 5500) |
| modules | C: tez-common tez-runtime-internals tez-runtime-library tez-dag tez-plugins/tez-history-parser tez-tools/analyzers/job-analyzer U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-95/6/console |
| versions | git=2.25.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by belugabehr** [comment]
@abstractdog Just a gentle reminder about this PR.

@pgaref Do you know anyone that can assist here?

**Comment by abstractdog** [comment]
hi @belugabehr! I just ran through the history, memory/speed improvement is promising, also you removed every questioned part, so this looks good to me
it's a +1 from my side

**Comment by belugabehr** [comment]
@abstractdog Can you please assist with a merge? Thanks!

**Comment by jteagles** [comment]
+1. Got distracted away from this PR. I'll do the merge shortly.

**Comment by belugabehr** [comment]
No worries! Thanks @jteagles

**Pull request merged**